app/models/provider.rb

```diff
@@ -42,8 +42,29 @@ class Provider < ApplicationRecord
   has_many :sites
   has_many :enrichments, foreign_key: :provider_code, primary_key: :provider_code, class_name: "ProviderEnrichment"
   def address_info
-    (enrichments.with_address_info.last || self)
-      .attributes_before_type_cast
-      .slice('address1', 'address2', 'address3', 'address4', 'postcode', 'region_code')
+    result = valid_enriched_address_info
+
+    if !result.present?
+      fields = %w[address1 address2 address3 address4 postcode]
+
+      result = self
+        .attributes_before_type_cast
+        .slice(*fields)
+    end
+
+    # region_code should be present in enrichment regardless
+    result['region_code'] = (enrichments.with_address_info.first || self).region_code_before_type_cast
+
+    result
+  end
+
+  def valid_enriched_address_info
+    fields = %w[address1 address2 address3 address4 postcode]
+
+    result = enrichments.with_address_info.first.present? && enrichments.with_address_info.first
```

Comment:
I think you can clean this up even further by using a functional style:

```ruby
return nil unless enrichments.with_address_info.present?

results = enrichments.with_address_info.first
                ...
```

Reply:
Alternatively:

```ruby
def valid_enriched_address_info
  fields = %w[address1 address2 address3 address4 postcode]

  return nil unless enrichments.with_address_info.present?

  enrichments.with_address_info.first
    .attributes_before_type_cast
    .slice(*fields)
    .select {|k,v| v.present? }
end
```

Reply:
cannot use select directly as its omits the field in question

Reply:
staled